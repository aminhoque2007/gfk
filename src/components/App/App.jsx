import React from 'react';

import SearchQuery from '../../components/SearchQuery';
import Styles from './App.scss';
import {loadStyles} from '../../utils';

const getClassName = loadStyles(Styles);

const App = () => (
  <main className={getClassName('Main')}>
    <SearchQuery />
  </main>
);

export default App;
