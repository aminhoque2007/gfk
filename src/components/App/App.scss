@import '../../stylesheets/colours';

.Main {
  background-color: $color-grey-light;
  font-size: 1em;
  height: 100%;
  max-width: 920px;
  padding: 30px;
  text-align: center;
}

@media screen and (min-width: 600px) {
  .Main {
    margin: 0 auto;
    width: 75%;
  }
}
