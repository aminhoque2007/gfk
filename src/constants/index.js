export const NETWORK_ERROR_MESSAGE
  = 'The request failed due to an unexpected error.';
export const NO_USER_RESULTS = 'No matching users. Try a different username.';
