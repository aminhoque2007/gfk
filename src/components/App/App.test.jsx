import React from 'react';
import {
  act,
  fireEvent,
  render,
  waitForElementToBeRemoved,
} from '@testing-library/react';
import { toHaveAttribute } from '@testing-library/jest-dom';
import { MockedProvider } from '@apollo/react-testing';

import App from './App';
import {
  mockWithError,
  mockWithNoResults,
  mockWithResults,
} from './__fixtures__/userQueriesResponse';
import { NETWORK_ERROR_MESSAGE, NO_USER_RESULTS } from '../../constants';

describe('App', () => {
  it('should render a text input and search button', () => {
    const { getByLabelText, getByDisplayValue } = render(<App />);
    expect(getByLabelText(/Username/)).toHaveAttribute('type', 'text');
    expect(getByDisplayValue('Search')).toHaveAttribute('type', 'submit');
  });

  it('should render 3 results when user clicks search button and only 3 matches returned by Github API', async () => {
    const { container, getByTestId } = render(
      <MockedProvider mocks={mockWithResults} addTypename={false}>
        <App />
      </MockedProvider>,
    );
    const form = getByTestId('searchForm');

    act(() => {
      fireEvent.submit(form, { target: { searchTerm: { value: 'miah' } } });
    });
    await waitForElementToBeRemoved(() => getByTestId('loading'));

    const spans = container.querySelectorAll('span');
    spans.forEach((span, index) =>
      expect(span.innerHTML).toEqual(
        mockWithResults[0].result.data.search.edges[index].node.login,
      ),
    );
  });

  it('should render a div with text stating no matching users were found when the Github API responds with no matches.', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mockWithNoResults} addTypename={false}>
        <App />
      </MockedProvider>,
    );
    const form = getByTestId('searchForm');

    act(() => {
      fireEvent.submit(form, {
        target: { searchTerm: { value: 'hlkadfhlajdsf' } },
      });
    });
    await waitForElementToBeRemoved(() => getByTestId('loading'));

    const usersNotFoundDiv = getByTestId('noMatchingUsers');
    expect(usersNotFoundDiv).toHaveTextContent(NO_USER_RESULTS);
  });

  it('should render a loading indicator while the App waits for a response from the Github API.', () => {
    const { getByTestId } = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <App />
      </MockedProvider>,
    );
    const form = getByTestId('searchForm');

    act(() => {
      fireEvent.submit(form, {
        target: { searchTerm: { value: 'hlkadfhlajdsf' } },
      });
    });

    const loadingDiv = getByTestId('loading');
    expect(loadingDiv).toMatchSnapshot();
  });

  it('should render an error message when the Github API responds with an error response.', async () => {
    const { getByTestId } = render(
      <MockedProvider mocks={mockWithError} addTypename={false}>
        <App />
      </MockedProvider>,
    );
    const form = getByTestId('searchForm');

    act(() => {
      fireEvent.submit(form, {
        target: { searchTerm: { value: 'hlkadfhlajdsf' } },
      });
    });
    await waitForElementToBeRemoved(() => getByTestId('loading'));

    const errorDiv = getByTestId('error');
    expect(errorDiv).toHaveTextContent(NETWORK_ERROR_MESSAGE);
  });
});
