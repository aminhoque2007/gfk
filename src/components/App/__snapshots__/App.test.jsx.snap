// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render a loading indicator while the App waits for a response from the Github API. 1`] = `
<div
  class="Search-query__loader"
  data-testid="loading"
/>
`;
